src/internal/kopia/base_finder.go

```diff
@@ -0,0 +1,259 @@
+			// backup as the base.
+			logger.CtxErr(ictx, err).Info("attempting to get snapshot")
+			continue
+		}
+
+		if len(man.IncompleteReason) > 0 {
+			if !foundIncomplete {
+				foundIncomplete = true
+
+				kopiaAssistSnaps = append(kopiaAssistSnaps, ManifestEntry{
+					Manifest: man,
+					Reasons:  []Reason{reason},
+				})
+			}
+
+			continue
+		}
+
+		// This is a complete snapshot so see if we have a backup model for it.
+		bup, err := b.getBackupModel(ictx, man)
+		if err != nil {
+			if errors.Is(err, data.ErrNotFound) {
+				continue
+			}
+
+			// Safe to continue here as we'll just end up attempting to use an older
+			// backup as the base.
+			logger.CtxErr(ictx, err).Debug("searching for base backup")
+
+			continue
+		}
+
+		ssid := bup.StreamStoreID
+		if len(ssid) == 0 {
+			ssid = bup.DetailsID
+		}
+
+		if len(ssid) == 0 {
+			continue
+		}
+
+		// If we've made it to this point then we're considering the backup
+		// complete as it has both an item data snapshot and a backup details
+		// snapshot.
+		me := ManifestEntry{
+			Manifest: man,
+			Reasons:  []Reason{reason},
+		}
+		kopiaAssistSnaps = append(kopiaAssistSnaps, me)
+
+		return &BackupEntry{
+			Backup:  bup,
+			Reasons: []Reason{reason},
+		}, &me, kopiaAssistSnaps, nil
+	}
+
+	logger.Ctx(ctx).Info("no base backups for reason")
+
+	return nil, nil, kopiaAssistSnaps, nil
+}
+
+func (b *baseFinder) getBase(
+	ctx context.Context,
+	reason Reason,
+	tags map[string]string,
+) (*BackupEntry, *ManifestEntry, []ManifestEntry, error) {
+	allTags := map[string]string{}
+
+	for _, k := range reason.TagKeys() {
+		allTags[k] = ""
+	}
+
+	maps.Copy(allTags, tags)
+	allTags = normalizeTagKVs(allTags)
+
+	metas, err := b.sm.FindManifests(ctx, allTags)
+	if err != nil {
+		return nil, nil, nil, clues.Wrap(err, "getting snapshots")
+	}
+
+	// No snapshots means no backups so we can just exit here.
+	if len(metas) == 0 {
+		return nil, nil, nil, nil
+	}
+
+	return b.findBasesInSet(ctx, reason, metas)
+}
+
+func (b *baseFinder) FindBases(
+	ctx context.Context,
+	reasons []Reason,
+	tags map[string]string,
+) (BackupBases, error) {
+	var (
+		// All maps go from ID -> entry. We need to track by ID so we can coalesce
+		// the reason for selecting something. Kopia assisted snapshots also use
+		// ManifestEntry so we have the reasons for selecting them to aid in
+		// debugging.
+		baseBups         = map[model.StableID]BackupEntry{}
+		baseSnaps        = map[manifest.ID]ManifestEntry{}
+		kopiaAssistSnaps = map[manifest.ID]ManifestEntry{}
+	)
+
+	for _, reason := range reasons {
+		ictx := clues.Add(
+			ctx,
+			"service", reason.Service.String(),
+			"category", reason.Category.String())
+		logger.Ctx(ictx).Info("searching for previous manifests")
+
+		baseBackup, baseSnap, assistSnaps, err := b.getBase(ictx, reason, tags)
+		if err != nil {
+			logger.Ctx(ctx).Info("error getting base snapshots for reason. Will fallback to full backup for it")
```

Comment:
imo "error doing foo" isn't a helpful message.  And shorter logs without sentence structure is more readable in log viewers.
```suggestion
			logger.Ctx(ctx).Info("getting base: falling back to full backup")
```